detector/standalone.go

```diff
@@ -1,48 +1,199 @@
 package detector
 
 import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strconv"
+	"sync"
+	"time"
+
 	log "github.com/golang/glog"
 	mesos "github.com/mesos/mesos-go/mesosproto"
+	"github.com/mesos/mesos-go/upid"
+	"golang.org/x/net/context"
+)
+
+const (
+	mesosHttpClientTimeout  = 10 * time.Second //TODO(jdef) configurable via fiag?
+	mesosLeaderSyncInterval = 30 * time.Second //TODO(jdef) configurable via fiag?
+	defaultMesosMasterPort  = 5050
 )
 
 type Standalone struct {
-	current *mesos.MasterInfo
-	ch      chan *mesos.MasterInfo
+	ch         chan *mesos.MasterInfo
+	client     *http.Client
+	tr         *http.Transport
+	pollOnce   sync.Once
+	initial    *mesos.MasterInfo
+	done       chan struct{}
+	cancelOnce sync.Once
 }
 
 // Create a new stand alone master detector.
 func NewStandalone(mi *mesos.MasterInfo) *Standalone {
-	return &Standalone{current: mi, ch: make(chan *mesos.MasterInfo)}
+	log.V(2).Infof("creating new standalone detector for %+v", mi)
+	ch := make(chan *mesos.MasterInfo)
+	tr := &http.Transport{}
+	return &Standalone{
+		ch: ch,
+		client: &http.Client{
+			Transport: tr,
+			Timeout:   mesosHttpClientTimeout,
+		},
+		tr:      tr,
+		initial: mi,
+		done:    make(chan struct{}),
+	}
 }
 
-// Trigger a master detected event.
-func (s *Standalone) Appoint(m *mesos.MasterInfo) {
-	log.V(2).Infoln("Appoint")
-	s.ch <- m
-	s.current = m
+func (s *Standalone) String() string {
+	return fmt.Sprintf("{initial: %+v}", s.initial)
 }
 
 // Detecting the new master.
 func (s *Standalone) Detect(o MasterChanged) error {
-	// go func() {
-	// 	for {
-	// 		receiver <- s.ch
-	// 		log.V(2).Infoln("Master detected")
-	// 	}
-	// }()
+	log.V(2).Info("Detect()")
+	s.pollOnce.Do(func() {
+		log.V(1).Info("spinning up asyc master detector poller")
+		go s.poller()
+	})
+	if o != nil {
+		log.V(1).Info("spawning asyc master detector listener")
+		go func() {
+			log.V(2).Infof("waiting for polled to send updates")
+		pollWaiter:
+			for {
+				select {
+				case mi, ok := <-s.ch:
+					if !ok {
+						break pollWaiter
+					}
+					log.V(1).Infof("detected master change: %+v", mi)
+					o.Notify(mi)
+				case <-s.done:
+					return
+				}
+			}
+			o.Notify(nil)
+		}()
+	} else {
+		log.Warningf("detect called with a nil master change listener")
+	}
 	return nil
 }
 
-// Stop the detection.
-func (s *Standalone) Stop() {
-	select {
-	case <-s.ch:
-		// already closed, don't close it again
-	default:
-		close(s.ch)
+func (s *Standalone) Done() <-chan struct{} {
+	return s.done
+}
+
+func (s *Standalone) Cancel() {
+	s.cancelOnce.Do(func() { close(s.done) })
+}
+
+// poll for changes to master leadership
```

Comment:
Maybe indicate the source of master leadership info, i.e. polls provided master for master leadership info via http://master:port/state.json REST service.